consensus/scheme/rolldpos/fsm.go

```diff
@@ -346,33 +345,40 @@ func (m *cFSM) handleStartRoundEvt(_ fsm.Event) (fsm.State, error) {
 		m.ctx.round = roundCtx{
 			height:          height,
 			number:          0,
-			timestamp:       m.ctx.clock.Now(),
 			endorsementSets: make(map[hash.Hash32B]*endorsement.Set),
-			proposer:        proposer,
 		}
 	}
-	if proposer == m.ctx.addr.RawAddress {
-		logger.Info().
-			Str("proposer", proposer).
-			Uint64("height", height).
-			Msg("current node is the proposer")
-		m.produce(m.newCEvt(eInitBlock), 0)
-		// TODO: we may need timeout event for block producer too
-		return sInitPropose, nil
-	}
-	logger.Info().
-		Str("proposer", proposer).
-		Uint64("height", height).
-		Msg("current node is not the proposer")
+	m.ctx.round.proposer = proposer
+	m.ctx.round.timestamp = m.ctx.clock.Now()
+
 	// Setup timeout for waiting for proposed block
+	m.produce(m.newCEvt(eInitBlock), 0)
 	m.produce(m.newTimeoutEvt(eProposeBlockTimeout), m.ctx.cfg.AcceptProposeTTL)
-	return sAcceptPropose, nil
+
+	return sInitPropose, nil
 }
 
 func (m *cFSM) handleInitBlockEvt(evt fsm.Event) (fsm.State, error) {
-	blk, err := m.ctx.mintBlock()
-	if err != nil {
-		return sEpochStart, errors.Wrap(err, "error when minting a block")
+	if m.ctx.round.proposer != m.ctx.addr.RawAddress {
+		logger.Info().
+			Str("proposer", m.ctx.round.proposer).
+			Uint64("height", m.ctx.round.height).
+			Uint32("round", m.ctx.round.number).
+			Msg("current node is not the proposer")
+		return sAcceptPropose, nil
+	}
+	logger.Info().
```

Comment:
Do you want to create it before L362, and only append different Msg in and out of the if condition

Reply:
will do